### ISSUE TITLE
Leadership election, local leader set to late?

### ISSUE THREAD

**Issue by philnate** [spring-cloud/spring-cloud-kubernetes#405]
Hi,

we migrated our Leader election code from Hazelcast to Kubernetes. The setup we have as of today is like:

```
@Override
    public void onApplicationEvent( AbstractLeaderEvent event )
    {
        boolean isLeader = event.getContext()
                                .isLeader();
        LOG.info( "Leadership changed, node is leader={}", isLeader );
        ...
     }
```

My expectation was that this will yield to the correct behaviour. Yet an `OnGrantedEvent`  where 
`event.getContext().isLeader()` is invoked will return `false`. On the contrary an `OnRevokedEvent` will yield for the same invocation `true`.

This is due the fact that LeadershipController is setting localLeader = newLeader after it notified on grant/revoke, which to me looks too late.

I can work around this issue by checking for the type of event received OnRevokedEvent/OnGrantedEvent rather than the isLeader() method. Yet I would assume that both should allow me to conclude the same result.

**Comment by geoand** [comment]
@gytis would you like to take a look?

**Comment by gytis** [comment]
@philnate this should fix it https://github.com/spring-cloud/spring-cloud-kubernetes/pull/409. Would you mind taking a look?

**Comment by gytis** [comment]
This bit in particular https://github.com/spring-cloud/spring-cloud-kubernetes/pull/409/files#diff-9f131b28928dd771bf1e8371048a0550. Othe changes are not releated, but without them there were issues with a sample app which I used to reproduce the issue.

**Comment by philnate** [comment]
@gytis looks good.

**Comment by ryanjbaxter** [comment]
Closed via f22175c
